_posts/2023-01-21-pthread-cancellation.markdown

```diff
@@ -0,0 +1,168 @@
+
+On the linux sysdeps, ``sys_before_cancellable_syscall`` is implemented as follows (code can be found [here](https://github.com/managarm/mlibc/blob/f2c6b8cd7fe67b89d12b23ba0a82e174ba2b7c0a/sysdeps/linux/generic/sysdeps.cpp#L365-L370)).
+
+```cpp
+int sys_before_cancellable_syscall(ucontext_t *uct) {
+	auto pc = reinterpret_cast<void*>(uct->uc_mcontext.rip);
+	if (pc < __mlibc_syscall_begin || pc > __mlibc_syscall_end)
+		return 0;
+	return 1;
+}
+```
+
+The function checks if the program counter of the thread is between the symbols ``__mlibc_syscall_begin`` and ``__mlibc_syscall_end``. These symbols are defined in the assembly routine for performing a syscall. On x86 that is as follows (code can be found [here](https://github.com/managarm/mlibc/blob/f2c6b8cd7fe67b89d12b23ba0a82e174ba2b7c0a/sysdeps/linux/x86_64/cp_syscall.S#L16-L23)).
+
+```x86asm
+__mlibc_syscall_begin:
+	/* tcbCancelEnableBit && tcbCancelTriggerBit */
+	and $((1 << 0) | (1 << 2)), %r11
+	cmp $((1 << 0) | (1 << 2)), %r11
+	je cancel
+	syscall
+__mlibc_syscall_end:
+	ret
+```
+
+The assembly routine checks if a thread should be canceled right before it performs the syscall. The labels allow mlibc to detect whether or not the syscall has been completed. If the interruption happens between the labels, then the syscall hasn't happened yet, no side effects have been produced, and the signal handler can continue as normal, which results in the cancellation of the thread.
+
+If however, the program counter is outside those labels, then mlibc cannot determine if the syscall has produced side-effects, so the cancellation is not performed within the signal handler, and the thread continues execution.
+
+Lastly, there is one more way a thread can get canceled, which is right _after_ a syscall has been performed. Below is an extract from mlibc's linux [syscall wrapper](https://github.com/managarm/mlibc/blob/ee4b75216b6b3276f89bdef168402a6c6bde0349/sysdeps/linux/generic/cxx-syscall.hpp#L163-L171).
+
+```cpp
+auto result = static_cast<sc_result_t>(do_nargs_cp_syscall(sc, sc_cast(args)...));
+if (int e = sc_error(result); e) {
+    auto tcb = reinterpret_cast<Tcb*>(get_current_tcb());
+    if (tcb_cancelled(tcb->cancelBits) && e == EINTR) {
+        __mlibc_do_cancel();
+        __builtin_unreachable();
+    }
+}
+return result;
+```
+
+If the syscall returns EINTR, it means that the syscall hasn't produced any side-effects, and so mlibc can cancel the thread if a cancellation is pending. There is one exception, which is for ``sys_close()``, where EINTR doesn't necessarily mean that no side-effects haven't been produced. Currently this is not handled in mlibc.
+
+[^1]: sysdeps are mlibc's terminology for describing functions that should be implemented by the target platform. This can be a syscall (e.g. ``sys_read()`` ), or functions that rely on specific details of the platform's syscall mechanisms.
```

Comment:
```suggestion
[^1]: sysdeps are mlibc's terminology for describing functions that should be implemented by the target platform. This can be a syscall (e.g. ``read(2)`` ), or functions that rely on specific details of the platform's syscall mechanisms.
```